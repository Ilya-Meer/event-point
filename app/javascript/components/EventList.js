import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import EventModal from '../components/EventModal';
import Event from '../components/Event';

const EventList = ({ events, updateEvents, user }) => {
  const [showProposeEvent, setShowProposeEvent] = useState(false);

  const renderEvents = () => {
    const formatted = events.map((event) => (
      <Event key={event.created_at} user={user} event={event} />
    ));
    return formatted;
  };

  return (
    <Fragment>
      <div className='eventlist-heading'>
        <h3>Proposed Events</h3>
        <Button onClick={() => setShowProposeEvent(true)}>
          + Propose Event
        </Button>
      </div>
      <div className='eventlist-wrapper'>{renderEvents()}</div>
      <EventModal
        show={showProposeEvent}
        handleDismiss={() => setShowProposeEvent(false)}
        user={user}
        events={events}
        updateEvents={updateEvents}
      />
    </Fragment>
  );
};

EventList.propTypes = {
  events: PropTypes.array,
  updateEvents: PropTypes.func,
  user: PropTypes.object,
};

EventList.defaultProps = {
  events: [],
  user: {},
};

export default EventList;
