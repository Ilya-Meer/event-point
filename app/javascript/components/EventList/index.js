import React, { Fragment, useState, useContext } from 'react';
import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import EventModal from '../EventModal';
import Event from '../Event';
import { deleteEvent } from '../../utils/api';
import { FlashContext } from '../../contexts/FlashContext';
import { SORTING_FILTERS, sortEvents } from '../../utils/sort';
import { isSuccessfulResponse, errorMessages } from '../../utils/error';

const EventList = ({ events, updateEvents, user }) => {
  const [sortingFilter, setSortingFilter] = useState(SORTING_FILTERS.date);
  const { setMessage } = useContext(FlashContext);

  const [eventToEdit, setEventToEdit] = useState({});
  const [showEventModal, setShowEventModal] = useState(false);

  const handleEditEvent = (id) => {
    const toEdit = events.find((event) => event.id === id);
    setEventToEdit(toEdit);
    setShowEventModal(true);
  };

  const handleDeleteEvent = async (id) => {
    try {
      const res = await deleteEvent(id);

      if (!isSuccessfulResponse(res)) {
        throw new Error(errorMessages.deleteEvent);
      }

      updateEvents(events.filter((event) => event.id !== id));
    } catch (error) {
      console.error(error);
      setMessage({ text: error.message, variant: 'danger' });
    }
  };

  const renderEvents = () => {
    const sorted = sortEvents(sortingFilter, events);

    const formatted = sorted.map((event) => (
      <Event
        key={event.id}
        user={user}
        event={event}
        handleEditEvent={handleEditEvent}
        handleDeleteEvent={handleDeleteEvent}
      />
    ));

    return formatted;
  };

  return (
    <Fragment>
      <div className='eventlist-heading'>
        <h3>Proposed Events</h3>
        <Button onClick={() => setShowEventModal(true)}>+ Propose Event</Button>
      </div>
      <div className='eventlist-wrapper'>{renderEvents()}</div>
      <EventModal
        eventToEdit={eventToEdit}
        show={showEventModal}
        handleDismiss={() => {
          setEventToEdit({});
          setShowEventModal(false);
        }}
        user={user}
        events={events}
        updateEvents={updateEvents}
      />
    </Fragment>
  );
};

EventList.propTypes = {
  events: PropTypes.array,
  updateEvents: PropTypes.func,
  user: PropTypes.object,
};

EventList.defaultProps = {
  events: [],
  user: {},
};

export default EventList;
