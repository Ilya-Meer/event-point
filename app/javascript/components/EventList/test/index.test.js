import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent, act } from '@testing-library/react';
import EventList from '..';

it('renders without crashing', () => {
  const tree = renderer
    .create(<EventList events={[]} updateEvents={() => {}} user={{}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders events properly', () => {
  const events = [
    {
      id: 1,
      created_at: new Date(new Date() - 10000).toISOString(),
      topic: 'Test Topic 1',
      description: 'Test Description 1',
      votes: [],
    },
    {
      id: 2,
      created_at: new Date().toISOString(),
      topic: 'Test Topic 2',
      description: 'Test Description 2',
      votes: [],
    },
  ];

  render(<EventList events={events} />);

  expect(screen.getAllByText(/suggested by/)).toHaveLength(2);
});

it('opens and closes the event submission modal dialog', () => {
  const events = [];
  const user = {};
  const updateEvents = () => {};

  render(<EventList events={events} user={user} updateEvents={updateEvents} />);

  expect(screen.queryByRole('dialog')).not.toBeInTheDocument();

  fireEvent.click(screen.getByText('+ Propose Event'));

  expect(screen.queryByRole('dialog')).toBeInTheDocument();
});

it('adds an event', async () => {
  const events = [];

  const user = {
    id: 5,
    email: 'test@test.com',
  };

  const updateEvents = jest.fn();

  render(<EventList events={events} user={user} updateEvents={updateEvents} />);

  // open modal to add an event
  fireEvent.click(screen.getByText('+ Propose Event'));

  const topicInput = screen.getByLabelText('Topic');
  const descriptionInput = screen.getByLabelText('Description');
  const submitEventButton = screen.getByText('Propose Event');

  const topic = 'An awesome topic';
  const description = 'A fascinating description';

  // fill out form in modal
  fireEvent.change(topicInput, { target: { value: topic } });
  fireEvent.change(descriptionInput, {
    target: { value: description },
  });

  fetchMock.once(JSON.stringify({ topic, description }));

  // submit event form
  await act(async () => {
    await fireEvent.click(submitEventButton);
  });

  expect(fetchMock).toHaveBeenCalledTimes(1);

  const apiPayload = {
    body: JSON.stringify({
      event: {
        topic,
        description,
        owner_id: user.id,
      },
    }),
    headers: { 'Content-Type': 'application/json' },
    method: 'POST',
  };

  // event has been created on the server
  expect(fetchMock).toHaveBeenCalledWith(
    'http://localhost:3000/events',
    apiPayload
  );

  // local state has been updated
  expect(updateEvents).toHaveBeenCalledWith([
    {
      topic,
      description,
      owner: 'test@test.com',
      votes: [],
    },
  ]);
});
