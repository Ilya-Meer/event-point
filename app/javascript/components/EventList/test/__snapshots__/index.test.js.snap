// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <div
    className="eventlist-heading"
  >
    <h3>
      Proposed Events
    </h3>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      + Propose Event
    </button>
  </div>,
  <div
    className="eventlist-wrapper"
  />,
]
`;
